### PR TITLE
osd: limit pg log length under all circumstances

### PR DESCRIPTION
**Pull request by neha-ojha**
Before this, `osd_max_pg_log_entries` was not a hard limit for the pg log length. This PR puts a hard cap on the number of log entries (and thus memory consumed) in the pg log, even during recovery and backfill.

Fixes: https://tracker.ceph.com/issues/23979
Signed-off-by: Neha Ojha <nojha@redhat.com>

### CODE REVIEW

**Comment by neha-ojha** [comment]
No related failures in rados run: http://pulpito.ceph.com/nojha-2018-07-17_03:31:35-rados-wip-23979-2018-07-16-distro-basic-smithi/

**Comment by neha-ojha** [comment]
retest this please

File: src/osd/PGLog.cc [hunk 1]
```diff
@@ -50,15 +50,8 @@ void PGLog::IndexedLog::trim(
...
-  if (complete_to != log.end() &&
-      complete_to->version <= s) {
-    generic_derr << " bad trim to " << s << " when complete_to is "
-		 << complete_to->version
-		 << " on " << *this << dendl;
-    assert(0 == "out of order trim");
-  }
-
   assert(s <= can_rollback_to);
+  generic_dout(0) << " complete_to " << complete_to->version << dendl;
```
Comment (jdurgin) [inline src/osd/PGLog.cc]: can use ldout(cct) instead of generic_dout - and should be higher level

Comment (neha-ojha) [reply src/osd/PGLog.cc]: the entire function uses generic_dout(), so left it the same for uniformity. Increased the level to 20.

**Comment by tchaikov** [comment]
retest this please.

**Comment by neha-ojha** [comment]
this is passing `make check` now, should be ready to merge.

**Pull request merged**